### ISSUE TITLE
Included NamedLock for interprocess and inter-isolate locking.

### ISSUE THREAD

**Issue by bsutton** [tekartik/synchronized.dart#9]
@alextekartik just thought I might bring to your attention some work I did in project I maintain called dshell.

It includes a class called 'NamedLock' which provides a heavy weight locking mechanism that provides locks between isolates and processes.

Normally I would issue a pull request with the code but for the foreseeable future I'm snowed under with other work.

But if you are interested in grabbing the code and including it in you package it would be a nice complement.

The code in question can be found here:
https://github.com/bsutton/dshell/blob/master/lib/src/util/named_lock.dart

**Comment by alextekartik** [comment]
Interesting. Initially synchronized was just about 'locking' in the same isolate (like java synchronized keyword). Cross isolate locking would be a welcome addition (especially in a flutter context). Cross process could also be interesting (more in a desktop context). Some people also asked me for cross tab (browser) locking, that's another story but by this I mean that having a similar API could help.

Some questions:
* Does this work on Flutter and have you tested on iOS and Android?
* I see in the code that you are hard-coding the UDP port. I guess at some point you need some shared reference but I wonder if it could conflict if some application are already using this port.
* Do you have performance information, i.e. how much time it takes to check whether it could enter in a 'locked' transaction. It is not always relevant (i.e. typically you only uses this for specific access) but always a nice to have information. Yes it depends on tons on things but comparing accessing a resource with and without a named lock could help.
* How did you come up with this implementation choice (UDP + file), is it something you have seen in another implementation/language or your own (smart) idea?
* System temp dir can (theorically) be cleared at any time by the system (although I never experienced that). Would this be an issue in your implementation?

**Comment by bsutton** [comment]
DShell is a cli only library and the NamedLock class was built with that in mind.

The clearing of the temp folder is an interesting issue but other than that I think it should work for for android/ios and flutter.

I've not done any performance testing but essentially given I don't believe there is any other way to achieve the same result its probably a moot point :)
For a normal cli app there has been is no noticeable performance overhead that would lead me to consider doing any testing but its probably a valid point.
As noted in my original post I do view this as a heavy weight lock but given we are talking about inter-process communications that's probably acceptable as generally this type of interactions don't do heavy levels of locking.

As to the technique, I've used ports as pseudo locks in other projects, but it was a very long time ago so I don't remember where the idea came from.
The hard coded port no. probably should be configurable. I did check documentation on well know port ranges so there should be a fairly minimal chance of a collision.  The problem with configuration is that it may be an issue that appears on certain devices but not others (based on certain apps not being installed). In reality there probably should be a process of 'hunting' for an available port and then that port no. written into a file. Of course this isn't guaranteed to work as some other app may start up after the hunting process completes. 
I think in reality it probably won't be an issue to the point I made the decision to ignore the issue :)
